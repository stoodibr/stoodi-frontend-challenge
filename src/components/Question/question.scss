#loading {
    height: 40vh;
    display: flex;
    justify-content: center;
    align-items: center;
}

#question-area {
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 2vh 0 !important;

    p {
        font-size: 1.8vh;
        font-weight: 500;
        color: var(--color-primary);
    }

}

.correctly {
    background-color: var(--color-light-success);
    color: var(--color-light-success);
    border-radius: 10px;
}

.error {
    background-color: var(--color-light-danger);
    color: var(--color-light-danger);
    border-radius: 10px;
}

#alternatives {
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 1vh 0 !important;

    #options {
        display: inline-block;
        width: 100%;
        height: auto;

        .option-area {
            height: 7vh;
            width: 100%;
            display: flex;
            align-items: center;
            justify-content: flex-start;
            margin: 1vh 0 !important;
            padding: 0 1vh !important;
        }

        label {
            display: flex;
            align-items: center;

            :hover {
                cursor: pointer;
            }

            :cu {
                color: var(--color-secundary);
            }
        }

        .text {
            color: var(--color-primary)
        }

        input[type="radio"] {
            width: 2.2vh;
            height: 2.2vh;
            margin: 1vh 2vh !important;
            color: var(--color-secundary);
        }

        input[type='radio']:disabled {
            pointer-events: none;
        }

        input:checked~.text {
            color: var(--color-secundary);
        }

    }
}

#response {
    width: 100%;
    padding: 1vh 0 !important;
    display: flex;
    justify-content: center;
    align-items: center;

    .correct-box {
        background-color: var(--color-light-success);
    }

    .incorrect-box {
        background-color: var(--color-light-danger);
    }

    #layout-size {
        border-radius: 4px;
        display: flex;
        width: 50%;
        padding: 2vh 2vh !important;
        justify-content: space-between;
        align-items: center;
        border-top: 1px solid var(--color-silver-light);
        border-bottom: 1px solid var(--color-silver-light);

        #result-text {
            display: inline-block;
            text-align: left;
            color: var(--color-primary);

            p:first-child {
                font-weight: 700;
            }

            p {
                padding: 0 !important;
                margin: 0 !important;
            }
        }
    }

    #btn-response {
        border-radius: 26px;
        border: none;
        background-color: var(--color-secundary);
        padding: 1.5vh 4vh !important;
        color: white;
        font-size: 1.6vh;
        font-weight: 400;
        outline: none;

        &:hover {
            cursor: pointer;
            background-color: #012A94;
        }

        &:disabled {
            background-color: #ccc;
            color: rgb(138, 138, 138);
            cursor: default;
        }
    }
}