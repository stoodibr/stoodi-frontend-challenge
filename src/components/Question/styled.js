import styled from 'styled-components';

export const StyledQuestion = styled.div`
  display: flex;
  flex-direction: column;
`;

export const QuestionTitle = styled.h1`
  font-size: 16px;
  margin-bottom: 20px;
`;

export const QuestionDescription = styled.div`
  font-size: 12px;
  margin-bottom: 20px;
`;

export const QuestionAlternativeArea = styled.div`
  margin-left: 5px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  margin-bottom: 20px;
`;

export const QuestionAlternative = styled.div`
  margin-left: 5px;
  font-size: 12px;
`;

export const ErrorMsg = styled.h1`
  font-size: 18px;
  color: #f7744a;
`;
