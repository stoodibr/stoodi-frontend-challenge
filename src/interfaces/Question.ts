export interface QuestionProps {
    exercise: {
        exercise_id: number,
        exercise_text: any,
        institution?: string,
        alternatives: [{
            letter: string,
            label: string
        }]
    }
}