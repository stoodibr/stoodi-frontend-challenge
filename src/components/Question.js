import React from "react";
import parse from "html-react-parser";
import styled from "styled-components";
import Radio from "@material-ui/core/Radio";
import AnswerBox from "./AnswerBox";
import { withStyles } from "@material-ui/core/styles";
import { green, orange, red } from "@material-ui/core/colors";

const CorrectRadio = withStyles({
  root: {
    color: green[600],
    "&$checked": {
      color: green[600],
    },
  },
  checked: {},
})((props) => <Radio color="default" {...props} />);

const WrongRadio = withStyles({
  root: {
    color: orange[800],
    "&$checked": {
      color: orange[800],
    },
  },
  checked: {},
})((props) => <Radio color="default" {...props} />);

const Container = styled.div`
  padding: 1rem;
  max-width: 768px;
`;
const Title = styled.h2`
  margin: 0px;
`;

const QuestionForm = styled.div`
  margin-top: 1.5rem;
`;
const AlternativesBox = styled.div`
  padding: 0.5rem 0rem;
  background-color: white;
`;

const RadioLabel = styled.span`
  margin-left: 0.5rem;
  font-size: 0.9rem;
`;
const RadioLetter = styled.span`
  margin-right: 0.7rem;
`;

const Question = ({ dados }) => {
  const [alternativa, setAlternativa] = React.useState("");
  const [choosed, setChoosed] = React.useState("");
  const [isCorrect, setIsCorrect] = React.useState(null);

  const handleCallBack = (isCorrect, choice) => {
    if (choice) {
      setIsCorrect(isCorrect);
      setChoosed(choice);
    } else {
      setIsCorrect(null);
      setChoosed("");
      setAlternativa(null);
    }
  };

  const radioBgColor = (letter) => {
    if (choosed !== letter) {
      return "white";
    } else if (isCorrect) {
      return green["100"];
    } else {
      return red["100"];
    }
  };

  return (
    <Container data-testid="question-container">
      <Title data-testid="question-title"> {dados.exercise.institution}</Title>
      {parse(dados.exercise.exercise_text.replace(":", `:<br/><br/>`))}
      <QuestionForm data-testid="question-form">
        {dados.exercise.alternatives.map((item) => {
          return (
            <AlternativesBox
              style={{ backgroundColor: radioBgColor(item.letter) }}
              key={item.letter}
            >
              <RadioLabel data-testid="question-radio-label">
                {choosed !== item.letter && (
                  <Radio
                    checked={alternativa === item.letter}
                    onChange={({ target }) => setAlternativa(target.value)}
                    value={item.letter}
                    color="primary"
                    name="radio-button"
                    inputProps={{ "aria-label": item.letter }}
                  />
                )}
                {choosed === item.letter && isCorrect && (
                  <CorrectRadio
                    checked={alternativa === item.letter}
                    onChange={({ target }) => setAlternativa(target.value)}
                    value={item.letter}
                    name="radio-button"
                    inputProps={{ "aria-label": item.letter }}
                  />
                )}
                {choosed === item.letter && !isCorrect && (
                  <WrongRadio
                    checked={alternativa === item.letter}
                    onChange={({ target }) => setAlternativa(target.value)}
                    value={item.letter}
                    name="radio-button"
                    inputProps={{ "aria-label": item.letter }}
                  />
                )}
                <RadioLetter>{`${item.letter}.`}</RadioLetter>
                {`${item.label}`}
              </RadioLabel>
            </AlternativesBox>
          );
        })}
        <AnswerBox
          alternativa={alternativa}
          cb={handleCallBack}
          id={dados.exercise.exercise_id}
        ></AnswerBox>
      </QuestionForm>
    </Container>
  );
};

export default Question;
