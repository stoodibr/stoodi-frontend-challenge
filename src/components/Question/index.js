import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ReactHtmlParser, {
  processNodes,
  convertNodeToElement,
  htmlparser2,
} from 'react-html-parser';
import getQuestion from '../../services/getQuestion';
import answerQuestion from '../../services/answerQuestion';
import * as S from './styled';
import Alternatives from '../Alternatives';
import QuestionButton from '../QuestionButton';

export default function Question() {
  const [newQuestion, setNewQuestion] = useState([]);
  const [newAnswer, setNewAnswer] = useState([]);
  const [answerResponse, setAnswerResponse] = useState({ is_correct: false });
  const [buttonState, setButtonState] = useState(false);
  const [answered, setAnswered] = useState(false);
  const [selectedAlt, setSelectedAlt] = useState('');
  const [errMessage, setErrMessage] = useState('');

  const activeButton = false;

  useEffect(async () => {
    try {
      setNewQuestion([await getQuestion()]);
    } catch (err) {
      setErrMessage(err.message);
    }
  }, []);

  useEffect(() => {
    if (newAnswer.length > 0) {
      setButtonState(true);
    }
  }, [newAnswer]);

  const getRadioValue = (value, questionId, alternativeId) => {
    setNewAnswer([{ questionId, value }]);
    setSelectedAlt(alternativeId);
  };

  const validateAnswer = async (value, questionId) => {
    try {
      const validate = await answerQuestion({
        questionId,
        questionChoice: value,
      });
      setAnswerResponse(validate);
      setAnswered(true);
    } catch (err) {
      setErrMessage(err.message);
    }
  };

  const nextAnswer = async () => {
    setNewAnswer([]);
    setAnswerResponse({ is_correct: false });
    setButtonState(false);
    setSelectedAlt('');
    setAnswered(false);
    setErrMessage('');
  };

  const resetAnswer = () => {
    setNewAnswer([]);
    setAnswerResponse({ is_correct: false });
    setButtonState(false);
    setSelectedAlt('');
    setAnswered(false);
    setErrMessage('');
  };

  return (
    <React.Fragment key="stoody">
      {errMessage ? (
        <S.ErrorMsg>
          Não foi possível executar a operação, erro: {errMessage}
        </S.ErrorMsg>
      ) : (
        newQuestion.map((ques) => (
          <S.StyledQuestion key={ques.exercise.exercise_id}>
            <S.QuestionTitle>
              <>{ques.exercise.institution}</>
            </S.QuestionTitle>
            <S.QuestionDescription>
              {ReactHtmlParser(ques.exercise.exercise_text)}
            </S.QuestionDescription>
            <S.QuestionAlternativeArea>
              <Alternatives
                alternatives={ques.exercise.alternatives}
                questionId={ques.exercise.exercise_id}
                getRadioValue={getRadioValue}
                key={ques.exercise.exercise_id}
                answered={answered}
                buttonState={buttonState}
                answerResponse={answerResponse.is_correct}
                newAnswer={newAnswer}
                selectedAlt={selectedAlt}
              />
            </S.QuestionAlternativeArea>
            <QuestionButton
              answerResponse={answerResponse.is_correct}
              buttonState={buttonState}
              validateAnswer={validateAnswer}
              newAnswer={newAnswer}
              answered={answered}
              nextAnswer={nextAnswer}
              resetAnswer={resetAnswer}
            />
          </S.StyledQuestion>
        ))
      )}
    </React.Fragment>
  );
}
