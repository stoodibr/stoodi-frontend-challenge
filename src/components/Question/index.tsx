import React, { useEffect, useState } from 'react'
import './question.scss'
import { Center } from '../style'
import { apiRequest } from '../../services/config-https'
import { QuestionProps } from '../../interfaces/Question'

export const Question = () => {
    const [question, setQuestion] = useState<QuestionProps>();
    const [textBtn, setTextBtn] = useState('');
    const [answerResult, setAnswerResult] = useState('');
    const [answered, setAnswered] = useState(false);
    const [loading, setLoading] = useState(false)
    const [loadingAwnser, setLoadingAwnser] = useState(false);
    const [optionSelected, setOptionSelected] = useState(false);
    const [chosenAnwser, setChosenAnwser] = useState('');
    const [disableOptions, setDisableOptions] = useState(false);
    const [statement, setStatement] = useState('');

    useEffect(() => {
        setAnswerResult('')
        setAnswered(false)
        setTextBtn('VERIFICAR RESPOSTA')
        setLoading(true)
        apiRequest.get(`dev/`)
            .then((response) => {
                const { data } = response;
                setQuestion(data)
                setLoading(false)
                let QuestionOriginal = data.exercise.exercise_text
                setStatement(QuestionOriginal.replace('<i>', '<br/> <i>'))
            })
            .catch((err) => {
                console.log(err);
                setLoading(false)
            });
    }, [])

    const selectOption = () => {
        setOptionSelected(true)
    }

    const responseQuestion = async () => {
        setLoadingAwnser(true)
        if (answered === true) {
            window.location.reload()
        } else {
            await apiRequest.post(`dev/`, {
                "exercise_id": question?.exercise.exercise_id,
                "choice": chosenAnwser
            })
                .then((response) => {
                    const { data } = response;
                    if (data.is_correct === true) {
                        setDisableOptions(true)
                        setAnswerResult('correct')
                        setTextBtn('PRÃ“XIMO')
                        let opcaoEscolhida = document.querySelector(`#${chosenAnwser}`)
                        opcaoEscolhida?.classList.add('correctly');
                        setAnswered(true)
                        setLoadingAwnser(false)

                    } else {
                        setDisableOptions(true)
                        setAnswerResult('incorrect')
                        setTextBtn('REFAZER')
                        let opcaoEscolhida = document.querySelector(`#${chosenAnwser}`)
                        opcaoEscolhida?.classList.add('error')
                        setAnswered(true)
                        setLoadingAwnser(false)
                    }
                })
                .catch((err) => {
                    console.log(err);
                });
        }
    }

    return (
        <>
            {loading ?
                <div id="loading">
                    <div className="spinner-border text-primary" role="status">
                        <span className="sr-only">Loading...</span>
                    </div>
                </div>

                :
                <>
                    <div id="question-area">
                        <Center>
                            <div>
                                <p style={{ textAlign: 'left' }}
                                    dangerouslySetInnerHTML={{ __html: statement }}
                                />
                            </div>

                        </Center>
                    </div>
                    <div id="alternatives">
                        <Center>
                            <article id='options'>
                                {question?.exercise.alternatives.map((object, key) => {
                                    return (
                                        <div className='option-area' id={`${object.letter}`} key={key}>
                                            <label>
                                                <input className="input-radio" disabled={disableOptions === true ? true : false} type="radio" value={object.letter} onChange={(e) => setChosenAnwser(object.letter)} onClick={(e) => selectOption()} name="light" />
                                                <span className="design"></span>
                                                <span className="text" id={`option${object.letter}`}>{object.letter}. {object.label}</span>
                                            </label>
                                        </div>
                                    )
                                })}

                            </article>
                        </Center>
                    </div>
                    <div id="response">
                        <div id="layout-size" className={answered === true ? answerResult === 'correct' ? 'correct-box' : 'incorrect-box' : ''}>
                            <div id="result-text" >
                                {loadingAwnser === true ?
                                    <div className="spinner-border text-primary" role="status">
                                        <span className="sr-only">Loading...</span>
                                    </div>
                                    : null}
                                {answered === true && loadingAwnser === false ?
                                    answerResult === 'correct' ?
                                        <>
                                            <p>Resposta correta</p>
                                            <p>Boa! Acertou em cheio.</p>
                                        </>
                                        :
                                        <>
                                            < p > Resposta incorreta</p>
                                            <p>Que tal tentar novamente?</p>
                                        </>
                                    : null}

                            </div>
                            <button type='button' disabled={optionSelected === true ? false : true} onClick={responseQuestion} id="btn-response">{textBtn}</button>
                        </div>
                    </div>
                </>
            }
        </>

    )
}