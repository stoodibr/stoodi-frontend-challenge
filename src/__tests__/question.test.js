import react from "react";
import { render, cleanup, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Question from "../components/Question";
import "jest-styled-components";
// import { rest } from "msw";
// import { setupServer } from "msw/node";
// TODO fix mocked API call
// const server = setupServer(
//   rest.get("/mockedReq", (req, res, ctx) => {
//     return res(
//       ctx.json({ exercise: { alternatives: [], institution: "Stoodi" } })
//     );
//   })
// );
// beforeAll(() => server.listen());
// afterEach(() => server.resetHandlers());
// afterAll(() => server.close());
// test("loads and displays data", async () => {
//   await render(<Question dados="/mockedReq" />);
// expect(screen.getByTestId("question-title")).toHaveTextContent(
//   "Stoodi"
// );
// });
afterEach(cleanup);

const hardMock = {
  exercise: {
    exercise_id: 3473,
    exercise_text:
      "<p>Assinale a alternativa que apresenta a raiz da equação abaixo: <i>2x - 7 = 9</i></p>",
    institution: "Stoodi",
    alternatives: [
      { letter: "a", label: "-8" },
      { letter: "b", label: "8" },
      { letter: "c", label: "10" },
      { letter: "d", label: "-10" },
    ],
  },
};

test("Question styled-components", async () => {
  render(<Question dados={hardMock} />);

  // Container styled component
  expect(screen.getByTestId("question-container")).toHaveStyle({
    padding: "1rem",
    maxWidth: "768px",
  });

  // Title styled component
  expect(screen.getByTestId("question-title")).toHaveTextContent("Stoodi");
  expect(screen.getByTestId("question-title")).toHaveStyle("margin: 0px");

  // QuestionForm styled component
  expect(screen.getByTestId("question-form")).toHaveStyle({
    marginTop: "1.5rem",
  });

  // QuestionForm styled component // ERROR due map generating elements with same ID
  // expect(screen.getByTestId("question-alternatives")).toHaveStyle({
  //   padding: "0.5rem 0rem",
  //   backgroundColor: "white",
  // });]
  expect(Question).toMatchSnapshot();
});
