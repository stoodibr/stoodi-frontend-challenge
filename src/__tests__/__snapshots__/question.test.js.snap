// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question styled-components 1`] = `[Function]`;
